### PR TITLE
Add SMB protocol support

### LINKED ISSUE #5368: Support for Samba (SMB) protocol
> **841fd6e** created the issue

Is there a possibility to add ability to access SMB (Samba) servers to CyberDuck?

The Finder implementation is spotty.

### PR DESCRIPTION
**Pull request by dkocher**
Fix #5368. In addition to #15009 this adds the following features

- [x] Allow configuration of domain in connection profile
- [x] [Add support to set timestamp on folders.](https://github.com/iterate-ch/cyberduck/pull/15020/commits/015c57415549b55aa5d0d8fb3a979327baa09945)
- [x] [Add quota feature implementation.](https://github.com/iterate-ch/cyberduck/pull/15020/commits/558976651aa95af67846ec4dec3e8efed5299318)
- [x] [List available shares on server using SMB2 IOCTL request.](https://github.com/iterate-ch/cyberduck/pull/15020/commits/fd9e8290b55eb316d0f9de1d1486eeb3e3d5b7da)

## Test Matrix

- [x] Windows 2016 Server 10.0.14393 `SMB_3_1_1`
- [x] Linux Samba `SMB_3_1_1`
- [x] macOS 13.5 (22G74) `SMB_3_0_2`

### CODE REVIEW

**Comment by dkocher** [comment]
Listing available shares when authenticated with SMB running on macOS we currently see the same error as described in [^1]  

[^1]: rapid7/smbj-rpc#129

File: core/src/main/java/ch/cyberduck/core/Scheme.java [hunk 1]
```diff
@@ -110,6 +110,17 @@ public int getPort() {
             return 443;
         }
     },
+    smb {
+        @Override
+        public boolean isSecure() {
+            return false;
```
Comment (ylangisc) [inline core/src/main/java/ch/cyberduck/core/Scheme.java:116]: Leads to the insecure connection alert? There is no secure alternative?

File: smb/src/main/java/ch/cyberduck/core/smb/SMBSession.java [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        catch(SMBRuntimeException e) {
+            throw new SMBExceptionMappingService().map("Cannot read container configuration", e);
+        }
+    }
+
+    public void releaseShare(final Path file) {
+        final String shareName = new SMBPathContainerService(this).getContainer(file).getName();
+        final ReentrantLock lock = locks.get(shareName);
+        if(null == lock) {
+            log.warn(String.format("Mising lock for %s", shareName));
```
Comment (ylangisc) [inline smb/src/main/java/ch/cyberduck/core/smb/SMBSession.java:156]: Typo.

File: core/src/main/java/ch/cyberduck/core/Scheme.java (hunk 1 above)
Comment (dkocher) [reply core/src/main/java/ch/cyberduck/core/Scheme.java:116]: Explicitly disabled in 953aa9b7. There is 
* SMB over QUIC (with TLS 1.3) that could be implemented.
* Windows Server 2022 and Windows 11 introduce AES-256-GCM and AES-256-CCM cryptographic suites for SMB 3.1.1 encryption [^1]. We could enable this in the dialect negotiation but it is not guaranteed the server does also support. As in this example with `SMB2_GLOBAL_CAP_ENCRYPTION`  from the integration tests

```
2023-08-24 22:48:31,727 [main] DEBUG ch.cyberduck.core.smb.SMBSession - Connected to ConnectionContext{
  serverGuid=34346430-6430-6438-3332-383300000000,
  serverName='localhost',
  negotiatedProtocol=NegotiatedProtocol{dialect=SMB_3_1_1, maxTransactSize=8388608, maxReadSize=8388608, maxWriteSize=8388608},
  clientGuid=379995c8-de62-4fcf-aa80-01af8a32de85,
  clientCapabilities=[SMB2_GLOBAL_CAP_DFS, SMB2_GLOBAL_CAP_LARGE_MTU, SMB2_GLOBAL_CAP_ENCRYPTION],
  serverCapabilities=[SMB2_GLOBAL_CAP_DFS, SMB2_GLOBAL_CAP_LEASING, SMB2_GLOBAL_CAP_LARGE_MTU],
  clientSecurityMode=1,
  serverSecurityMode=1,
  server='com.hierynomus.smbj.server.Server@32a2a6be'
}
```
 
[^1]: https://learn.microsoft.com/en-us/windows-server/storage/file-server/smb-security#smb-encryption

**Review by ylangisc — APPROVED** [review]

File: profiles/default/SMB.cyberduckprofile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        <key>Path Configurable</key>
+        <true/>
+        <key>Hostname Configurable</key>
+        <true/>
+        <key>Username Configurable</key>
+        <true/>
+        <key>Password Configurable</key>
+        <true/>
+        <key>Anonymous Configurable</key>
+        <true/>
```
Comment (AliveDevil) [inline profiles/default/SMB.cyberduckprofile:37]: Upstream blocks anonymous configuration. Disable.

**Review by AliveDevil — CHANGES_REQUESTED** [review]

File: profiles/default/SMB.cyberduckprofile (hunk 3 above)
Comment (AliveDevil) [reply profiles/default/SMB.cyberduckprofile:37]: Approving when anonymous disabled: #15069

**Comment by AliveDevil** [comment]
UPN (`user@domain.com`) is not supported by NTLM, requires Negotiate/Kerberos - thus support isn't possible.

**Comment by AliveDevil** [comment]
Decided to post-pone Windows Integrated Authentication/SSO - instructions unclear.

**Review by AliveDevil — APPROVED** [review]

**Pull request merged**